activemq-broker/src/main/java/org/apache/activemq/broker/TransactionBroker.java

```diff
@@ -291,13 +294,39 @@ public void send(ProducerBrokerExchange producerExchange, final Message message)
             transaction = getTransaction(context, message.getTransactionId(), false);
         }
         context.setTransaction(transaction);
+
         try {
+            // [AMQ-9344] Limit uncommitted transactions by count
+            verifyUncommittedCount(producerExchange, transaction, message);
             next.send(producerExchange, message);
         } finally {
             context.setTransaction(originalTx);
         }
     }
 
+    protected void verifyUncommittedCount(ProducerBrokerExchange producerExchange, Transaction transaction, Message message) throws Exception {
+        // maxUncommittedCount <= 0 disables
+        int maxUncommittedCount = this.getBrokerService().getMaxUncommittedCount();
+        if (maxUncommittedCount > 0 && transaction.size() >= maxUncommittedCount) {
```

Comment:
Shouldn't this be a greater than check and not >= ? We only want to fail if we go past the max size not if we are at it. This is why the test has 11 failed ops but should only be 10 since your test you set the max to 10 so we should allow 10 and then fail on 11 - 20

Reply:
Nevermind, equals makes sense since you've already sent the max previous (this is on the send of the next one) so i'm still wondering why the test assertion is 11 and not 10 though.